lib/puppet/pops/puppet_stack.rb

```diff
@@ -1,57 +1,62 @@
-module Puppet::Pops
-# Module for making a call such that there is an identifiable entry on
-# the ruby call stack enabling getting a puppet call stack
-# To use this make a call with:
-# ```
-# Puppet::Pops::PuppetStack.stack(file, line, receiver, message, args)
-# ```
-# To get the stack call:
-# ```
-# Puppet::Pops::PuppetStack.stacktrace
-#
-# When getting a backtrace in Ruby, the puppet stack frames are
-# identified as coming from "in 'stack'" and having a ".pp" file
-# name.
-# To support testing, a given file that is an empty string, or nil
-# as well as a nil line number are supported. Such stack frames
-# will be represented with the text `unknown` and `0´ respectively.
-#
-module PuppetStack
-  # Pattern matching an entry in the ruby stack that is a puppet entry
-  PP_ENTRY_PATTERN = /^(.*\.pp)?:([0-9]+):in (`stack'|`block in call_function'|`<eval>')/
+require 'puppet/thread_local'
 
-  # Sends a message to an obj such that it appears to come from
-  # file, line when calling stacktrace.
-  #
-  def self.stack(file, line, obj, message, args, &block)
-    file = '' if file.nil?
-    line = 0 if line.nil?
+module Puppet
+  module Pops
+    # Utility class for keeping track of the "Puppet stack", ie the file
+    # and line numbers of Puppet Code that created the current context.
+    #
+    # To use this make a call with:
+    #
+    # ```rb
+    # Puppet::Pops::PuppetStack.stack(file, line, receiver, message, args)
+    # ```
+    #
+    # To get the stack call:
+    #
+    # ```rb
+    # Puppet::Pops::PuppetStack.stacktrace
+    # ```
+    #
+    # or
+    #
+    # ```rb
+    # Puppet::Pops::PuppetStack.top_of_stack
+    # ```
+    #
+    # To support testing, a given file that is an empty string, or nil
+    # as well as a nil line number are supported. Such stack frames
+    # will be represented with the text `unknown` and `0´ respectively.
+    class PuppetStack
```

Comment:
This was a module before...

I checked in our code base and stdlib and only found usages of it that called the module level methods directly on the module (which is also how we document its api), rather than including the module into another class.

This would break users who were including it, I don't know if we have any idea if any users are including it into other classes though....

Reply:
Moved this to be a module and it seems fine.